packages/babel-parser/test/fixtures/typescript/types/infer-with-constraints-babel-7/output.json

```diff
@@ -0,0 +1,442 @@
+{
+  "type": "File",
+  "start":0,"end":426,"loc":{"start":{"line":1,"column":0,"index":0},"end":{"line":5,"column":84,"index":426}},
+  "program": {
+    "type": "Program",
+    "start":0,"end":426,"loc":{"start":{"line":1,"column":0,"index":0},"end":{"line":5,"column":84,"index":426}},
+    "sourceType": "module",
+    "interpreter": null,
+    "body": [
+      {
+        "type": "TSTypeAliasDeclaration",
+        "start":0,"end":74,"loc":{"start":{"line":1,"column":0,"index":0},"end":{"line":1,"column":74,"index":74}},
+        "id": {
+          "type": "Identifier",
+          "start":5,"end":7,"loc":{"start":{"line":1,"column":5,"index":5},"end":{"line":1,"column":7,"index":7},"identifierName":"X3"},
+          "name": "X3"
+        },
+        "typeParameters": {
+          "type": "TSTypeParameterDeclaration",
+          "start":7,"end":10,"loc":{"start":{"line":1,"column":7,"index":7},"end":{"line":1,"column":10,"index":10}},
+          "params": [
+            {
+              "type": "TSTypeParameter",
+              "start":8,"end":9,"loc":{"start":{"line":1,"column":8,"index":8},"end":{"line":1,"column":9,"index":9}},
+              "name": "T"
+            }
+          ]
+        },
+        "typeAnnotation": {
+          "type": "TSConditionalType",
+          "start":13,"end":73,"loc":{"start":{"line":1,"column":13,"index":13},"end":{"line":1,"column":73,"index":73}},
+          "checkType": {
+            "type": "TSTypeReference",
+            "start":13,"end":14,"loc":{"start":{"line":1,"column":13,"index":13},"end":{"line":1,"column":14,"index":14}},
+            "typeName": {
+              "type": "Identifier",
+              "start":13,"end":14,"loc":{"start":{"line":1,"column":13,"index":13},"end":{"line":1,"column":14,"index":14},"identifierName":"T"},
+              "name": "T"
+            }
+          },
+          "extendsType": {
+            "type": "TSTupleType",
+            "start":23,"end":47,"loc":{"start":{"line":1,"column":23,"index":23},"end":{"line":1,"column":47,"index":47}},
+            "elementTypes": [
+              {
+                "type": "TSInferType",
+                "start":24,"end":46,"loc":{"start":{"line":1,"column":24,"index":24},"end":{"line":1,"column":46,"index":46}},
+                "typeParameter": {
+                  "type": "TSTypeParameter",
+                  "start":30,"end":46,"loc":{"start":{"line":1,"column":30,"index":30},"end":{"line":1,"column":46,"index":46}},
+                  "name": "U",
+                  "constraint": {
```

Comment:
I am really surprised by this placement. I'd expect `constraint` to be on `TSInferType`, not on `TSTypeParameter`.

Reply:
I think instead of adding a new `constraint` field to `TSInferType`, we should add a `constraint` to the `typeParamter` of `TSInferType`. Because it is the type parameter, not the `infer`, that is constrained.

FYI, the AST in TypeScript Compiler adds `constaint` to `TypeParameter`. ([TypeScript AST Viewer Link](https://ts-ast-viewer.com/#code/C4TwDgpgBAGgPAVQDRQCoD4oF41QgD2AgDsATAZygEtiAzCAJygT0JIqmIFcBbAI0ZQA-MygAuNAG4AUEA))

Reply:
Tbh, I really dislike many choices that the TS team made with their AST :stuck_out_tongue: 

However, I realized that your AST design follows the shape we already have for `type X<A extends B>` so it's fine.